File: src/main/scala/com/kirkwoodwest/scalawoods/DoubleGrid.scala
package com.kirkwoodwest.scalawoods

import com.bitwig.extension.controller.api.*
import com.kirkwoodwest.scalawoods.GridButtonState

import scala.collection.mutable

class DoubleGrid(
                  val host: ControllerHost,
                  val id: String,
                  val cursorTrackList: List[CursorTrack],
                  val hardware: HardwareLaunchPadMiniMK3
                ) {
  val clipLauncherSlotBanks: Seq[ClipLauncherSlotBank] = cursorTrackList.map { cursorTrack =>
    cursorTrack.clipLauncherSlotBank
  }

  val sceneBank: SceneBank = host.createSceneBank(4)

  var launchSceneModeActive: Boolean = false
  val clipButtonActions: Array[Array[mutable.HashMap[ActionID, HardwareActionBindable]]] =
    Array.ofDim[mutable.HashMap[ActionID, HardwareActionBindable]](8, 8)

  for {
    i <- clipButtonActions.indices
    j <- clipButtonActions(i).indices
  } {
    clipButtonActions(i)(j) = mutable.HashMap.empty[ActionID, HardwareActionBindable]
  }

  var altLaunchModeActive: Boolean = false
  var stopModeActive: Boolean = false
  val transport: Transport = host.createTransport
  transport.isClipLauncherOverdubEnabled.markInterested()
  transport.isClipLauncherAutomationWriteEnabled.markInterested()

  // Setup Fill Mode
  hardware
    .getSceneButton(0)
    .button
    .pressedAction
    .addBinding(transport.isFillModeActive.setToTrueAction())
  hardware
    .getSceneButton(0)
    .button
    .releasedAction
    .addBinding(transport.isFillModeActive.setToFalseAction())
  hardware.getSceneButton(ButtonIndexes.FillActive).setDirty()

  //Fill Led
  transport.isFillModeActive.addValueObserver((active: Boolean) => {
    if (active) {
      hardware.getSceneButton(0).setState(GridButtonState.ClipPlaying)
    } else {
      hardware.getSceneButton(0).setState(GridButtonState.HasClip)
    }
  })

  // Alternate Launch Mode
  val launchModeToggle: HardwareActionBindable = host.createAction(
    () => {
      toggleLaunchModeState()
    },
    () => "Alt Launch Mode Active"
  )

  // Setup buttons for moving up and down the grid
  val clipLauncherUp: HardwareActionBindable = host.createAction(
    () => {
      moveClipLauncher(-1)
    },
    () => "Move Grid Up"
  )
  val clipLauncherDown: HardwareActionBindable = host.createAction(
    () => {
      moveClipLauncher(1)
    },
    () => "Move Grid Down"
  )

  hardware.getFuncButton(0).button.pressedAction().addBinding(clipLauncherUp)
  hardware.getFuncButton(1).button.pressedAction().addBinding(clipLauncherDown)

  //Scene Mode
  val launchSceneActiveAction: HardwareActionBindable = host.createAction(
    () => {
      setSceneLaunchMode(true)
    },
    () => "LaunchScene Mode Active"
  )
  val launchSceneInactiveAction: HardwareActionBindable = host.createAction(
    () => {
       setSceneLaunchMode(false)
    },
    () => "LaunchScene Mode Deactive"
  )
  hardware.getSceneButton(ButtonIndexes.LaunchScene).button.pressedAction().addBinding(launchSceneActiveAction)
  hardware.getSceneButton(ButtonIndexes.LaunchScene).button.releasedAction().addBinding(launchSceneInactiveAction)

  //Stop Mode
  val stopModeActiveAction: HardwareActionBindable = host.createAction(
    () => {
      setStopMode(true)
    },
    () => "Stop Mode Active"
  )
  val stopModeInactiveAction: HardwareActionBindable = host.createAction(
    () => {
      setStopMode(false)
    },
    () => "Stop Mode Inactive"
  )
  hardware.getSceneButton(ButtonIndexes.Stop).button.pressedAction().addBinding(stopModeActiveAction)
  hardware.getSceneButton(ButtonIndexes.Stop).button.releasedAction().addBinding(stopModeInactiveAction)

  // Loop through cursor track list and build out the clip launcher slots
  createButtonBindings()
  updateState()



  private def setStopMode(bool: Boolean): Unit =
    stopModeActive = bool
    updateState()

  private def setSceneLaunchMode(active:Boolean): Unit =
    launchSceneModeActive = active
    updateState()

  private def updateState(): Unit = {
    if(stopModeActive) {
      hardware.getSceneButton(ButtonIndexes.Stop).setState(GridButtonState.NoteRecording)
    } else {
      hardware.getSceneButton(ButtonIndexes.Stop).setState(GridButtonState.ClipPlaying)
    }
    if (launchSceneModeActive && !stopModeActive) {
      hardware.getSceneButton(ButtonIndexes.LaunchScene).setState(GridButtonState.NoteRecording)

      if (altLaunchModeActive) {
        bindSceneAltButtonsClips()
        hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonState.NoteRecording)
      } else {
        bindSceneButtonsClips()
        hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonState.ClipPlaying)
      }
    } else if (stopModeActive) {
        bindStopButtons()
        hardware.getSceneButton(ButtonIndexes.Stop).setState(GridButtonState.ClipPlaying)
    } else {
      hardware.getSceneButton(ButtonIndexes.LaunchScene).setState(GridButtonState.HasClip)
      bindGridButtonsClips()
      if(altLaunchModeActive) {
        bindGridAltButtonsClips()
        hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonState.NoteRecording)
      } else {
        bindGridButtonsClips()
        hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonState.ClipPlaying)
      }
      hardware.getSceneButton(ButtonIndexes.FillActive).setState(GridButtonState.HasClip)
    }
  }
  private def toggleLaunchModeState(): Unit = {
    // Implement the state toggling logic here
    altLaunchModeActive = !altLaunchModeActive
    updateState()
  }

  
  private def buttonMap(col: Int, row: Int): (Int, Int) = {
    var r = row
    if (col >= 8) {
      r = row + 4
    }
    (col % 8, r)
  }

  private def createButtonBindings(): Unit = {
    clipLauncherSlotBanks.zipWithIndex.foreach { (clipLauncherSlotBank, trackIndex) =>
      clipLauncherSlotBank.scrollPosition().markInterested()
      val sizeOfBank = clipLauncherSlotBank.getSizeOfBank

      for (row_index <- 0 until sizeOfBank) {
        val (c, r) = buttonMap(trackIndex, row_index)
        val clipLauncherSlot: ClipLauncherSlot = clipLauncherSlotBank.getItemAt(row_index)
        val launchAction: HardwareActionBindable = clipLauncherSlot.launchAction()
        val launchReleaseAction: HardwareActionBindable = clipLauncherSlot.launchReleaseAction()
        val launchAltAction: HardwareActionBindable = clipLauncherSlot.launchAltAction()
        val launchReleaseAltAction: HardwareActionBindable = clipLauncherSlot.launchReleaseAltAction()
        val stopAction: HardwareActionBindable = clipLauncherSlotBank.stopAction()
        val stopAltAction: HardwareActionBindable = clipLauncherSlotBank.stopAltAction()

        val sceneLaunchAction: HardwareActionBindable = sceneBank.getItemAt(row_index).launchAction()
        val sceneAltLaunchAction: HardwareActionBindable = sceneBank.getItemAt(row_index).launchAltAction()
        val sceneReleaseAction: HardwareActionBindable = sceneBank.getItemAt(row_index).launchReleaseAction()
        val sceneAltReleaseAction: HardwareActionBindable = sceneBank.getItemAt(row_index).launchReleaseAltAction()

        val actions = mutable.HashMap[ActionID, HardwareActionBindable](
          ActionID.ClipLaunch -> launchAction,
          ActionID.ClipLaunchRelease -> launchReleaseAction,
          ActionID.ClipAltLaunch -> launchAltAction,
          ActionID.ClipAltLaunchRelease -> launchReleaseAltAction,
          ActionID.ClipStop -> stopAction,
          ActionID.SceneLaunch -> sceneLaunchAction,
          ActionID.SceneAltLaunch -> sceneAltLaunchAction,
          ActionID.SceneLaunchRelease -> sceneReleaseAction,
          ActionID.SceneAltLaunchRelease -> sceneAltReleaseAction,
        )

        clipButtonActions(c)(r) = actions

        val gridButton: GridButton = hardware.getGridButton(c, r)
        val gridStatus = GridStatus(gridButton, clipLauncherSlot, cursorTrackList(trackIndex))
      }
    }
  }

  private def bindGridButtonsClips(): Unit = {
    bindActionsToGrid(ActionID.ClipLaunch, ActionID.ClipLaunchRelease)
  }

  private def bindGridAltButtonsClips(): Unit = {
    bindActionsToGrid(ActionID.ClipAltLaunch, ActionID.ClipAltLaunchRelease)
  }

  private def bindSceneButtonsClips(): Unit = {
    bindActionsToGrid(ActionID.SceneLaunch, ActionID.SceneLaunchRelease)
  }

  private def bindSceneAltButtonsClips(): Unit = {
    bindActionsToGrid(ActionID.SceneAltLaunch, ActionID.SceneAltLaunchRelease)
  }

  private def bindStopButtons(): Unit = {
    bindActionsToGrid(ActionID.ClipStop, null)
  }
  private def bindStopAltButtons(): Unit = {
    bindActionsToGrid(ActionID.ClipAltStop, null)
  }

  private def bindActionsToGrid(pressedAction: ActionID, releasedAction: ActionID): Unit = {
    for (col <- 0 until 8) {
      for (row <- 0 until 8) {
        bindActionsToGridButton(col, row, pressedAction, releasedAction)
      }
    }
  }

  private def bindActionsToGridButton(col:Int, row:Int, pressedAction:ActionID, releasedAction:ActionID): Unit = {
    val actions = clipButtonActions(col)(row)
    if(pressedAction != null)
      val action = actions(pressedAction)
      hardware.getGridButton(col, row).button.pressedAction().setBinding(action)
    else
      hardware.getGridButton(col, row).button.pressedAction().clearBindings()
    if (releasedAction != null)
      val action = actions(releasedAction)
      hardware.getGridButton(col, row).button.releasedAction().setBinding(action)
    else
      hardware.getGridButton(col, row).button.releasedAction().clearBindings()
  }

  def moveClipLauncher(direction: Int): Unit = {
    val steps: Int = direction * 4
    cursorTrackList.zipWithIndex.foreach { case (cursorTrack, trackIndex) =>
      val clipLauncherSlotBank: ClipLauncherSlotBank = cursorTrack.clipLauncherSlotBank
      val index = clipLauncherSlotBank.scrollPosition().get
      val sizeOfBank = clipLauncherSlotBank.getSizeOfBank
      clipLauncherSlotBank.scrollPosition().set(index + steps)
    }
  }
}

//Scene Button Indexes
private object ButtonIndexes {
  val FillActive: Int = 0
  val FillAlt: Int = 1
  val LaunchMode: Int = 2
  val LaunchScene: Int = 3
  val Record: Int = 4
  val Stop: Int = 5
}

enum Modes:
  case  Clip,
        ClipAlt,
        Scene,
        SceneAlt

enum ActionID:
  case  ClipLaunch,
        ClipLaunchRelease,
        ClipAltLaunch,
        ClipAltLaunchRelease,
        SceneLaunch,
        SceneLaunchRelease,
        SceneAltLaunch,
        SceneAltLaunchRelease,
        ClipStop,
        ClipAltStop

File: src/main/java/com/kirkwoodwest/launchpadminimk3/DoubleGrid.java
package com.kirkwoodwest.launchpadminimk3;

import com.bitwig.extension.controller.api.*;
import com.kirkwoodwest.launchpadminimk3.hardware.GridButton;
import com.kirkwoodwest.launchpadminimk3.hardware.GridButtonColor;
import com.kirkwoodwest.launchpadminimk3.hardware.HardwareLaunchPadMiniMK3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DoubleGrid {
  private final ControllerHost host;
  private final String id;
  private final List<CursorTrack> cursorTrackList;
  private final HardwareLaunchPadMiniMK3 hardware;
  private final List<ClipLauncherSlotBank> clipLauncherSlotBanks = new ArrayList<>();
  private final SceneLauncher sceneLauncher;
  private final Fill fill;
  private final SceneBank sceneBank;
  private boolean launchSceneModeActive;
  private final Map<ActionID, HardwareActionBindable>[][] clipButtonActions;
  private ArrayList<ArrayList<GridStatus>> gridStatus;
  private DoubleGridState doubleGridState;
  private boolean altLaunchModeActive;
  private boolean stopModeActive;
  private boolean recordModeActive;
  private final Transport transport;
  private boolean deleteModeActive;


  public DoubleGrid(ControllerHost host, String id, List<CursorTrack> cursorTrackList, HardwareLaunchPadMiniMK3 hardware) {
    this.host = host;
    this.id = id;
    this.cursorTrackList = cursorTrackList;
    this.hardware = hardware;

    //Indicate where the scenes are
    this.sceneBank = host.createSceneBank(4);
    this.sceneBank.scrollPosition().markInterested();
    sceneBank.setIndication(true);

    //Fill
    fill = new Fill(host, host.createTransport());

    for (int i = 0; i < cursorTrackList.size(); i++) {
      ClipLauncherSlotBank clipLauncherSlotBank = cursorTrackList.get(i).clipLauncherSlotBank();
      clipLauncherSlotBanks.add(clipLauncherSlotBank);
    }

    this.launchSceneModeActive = false;
    this.clipButtonActions = new HashMap[8][8];

    for (int i = 0; i < 8; i++) {
      for (int j = 0; j < 8; j++) {
        this.clipButtonActions[i][j] = new HashMap<>();
      }
    }


    this.altLaunchModeActive = false;
    this.stopModeActive = false;
    this.transport = host.createTransport();
    this.transport.isClipLauncherOverdubEnabled().markInterested();
    this.transport.isClipLauncherAutomationWriteEnabled().markInterested();

    //Create Scene Launcher
    sceneLauncher = new SceneLauncher(host, 4, clipLauncherSlotBanks);
    setupHardware();
    //Create Scene Launcher

    createButtonBindings();
    updateState();
  }

  private void setupHardware() {
    hardware.getSceneButton(ButtonIndexes.FillActive).getButton().pressedAction().addBinding(fill.getFillPressedAction());
    hardware.getSceneButton(ButtonIndexes.FillActive).getButton().releasedAction().addBinding(fill.getFillReleasedAction());
    hardware.getSceneButton(ButtonIndexes.FillActive).setDirty();
    hardware.getSceneButton(ButtonIndexes.FillAlt).getButton().pressedAction().addBinding(fill.getFillOptionPressedAction());
    hardware.getSceneButton(ButtonIndexes.FillAlt).setDirty();

    Runnable updateLed = () -> {
      boolean option = this.fill.isFillOptionActive().get();
      boolean fill = this.fill.isFillModeActive().get();
      boolean locked = this.fill.isFillModeLocked().get();

      if (option){
        hardware.getSceneButton(ButtonIndexes.FillActive).setState(GridButtonColor.FillModeBar);
        hardware.getSceneButton(ButtonIndexes.FillAlt).setState(GridButtonColor.FillModeBar);
      } else if ( locked) {
        hardware.getSceneButton(ButtonIndexes.FillActive).setState(GridButtonColor.FillModeLocked);
        hardware.getSceneButton(ButtonIndexes.FillAlt).setState(GridButtonColor.FillModeLocked);
      } else if (fill) {
        hardware.getSceneButton(ButtonIndexes.FillActive).setState(GridButtonColor.FillModeActive);
        hardware.getSceneButton(ButtonIndexes.FillAlt).setState(GridButtonColor.FillModeInactive);
      } else {
        hardware.getSceneButton(ButtonIndexes.FillActive).setState(GridButtonColor.FillModeInactive);
        hardware.getSceneButton(ButtonIndexes.FillAlt).setState(GridButtonColor.FillModeInactive);
      }
    };

    fill.isFillModeActive().addValueObserver(active -> {
      updateLed.run();
    });

    fill.isFillOptionActive().addValueObserver(active -> {
      updateLed.run();
    });

    fill.isFillModeLocked().addValueObserver(active -> {
      updateLed.run();
    });

    //Move Clip Launcher Grid
    HardwareActionBindable clipLauncherUp = host.createAction(() -> moveClipLauncher(-1), () -> "Move Grid Up");
    HardwareActionBindable clipLauncherDown = host.createAction(() -> moveClipLauncher(1), () -> "Move Grid Down");

    hardware.getFuncButton(0).getButton().pressedAction().addBinding(clipLauncherUp);
    hardware.getFuncButton(1).getButton().pressedAction().addBinding(clipLauncherDown);

    //Launch Mode Toggle
    HardwareActionBindable launchModeToggle = host.createAction(() -> toggleLaunchModeState(), () -> "Alt Launch Mode Active");
    hardware.getSceneButton(ButtonIndexes.LaunchMode).getButton().pressedAction().addBinding(launchModeToggle);

    HardwareActionBindable launchSceneActiveAction = host.createAction(() -> setSceneLaunchMode(true), () -> "LaunchScene Mode Active");
    HardwareActionBindable launchSceneInactiveAction = host.createAction(() -> setSceneLaunchMode(false), () -> "LaunchScene Mode Deactive");
    hardware.getSceneButton(ButtonIndexes.LaunchScene).getButton().pressedAction().addBinding(launchSceneActiveAction);
    hardware.getSceneButton(ButtonIndexes.LaunchScene).getButton().releasedAction().addBinding(launchSceneInactiveAction);

    //Stop Button
    HardwareActionBindable stopModeActiveAction = host.createAction(() -> setStopMode(true), () -> "Stop Mode Active");
    HardwareActionBindable stopModeInactiveAction = host.createAction(() -> setStopMode(false), () -> "Stop Mode Inactive");
    hardware.getSceneButton(ButtonIndexes.Stop).getButton().pressedAction().addBinding(stopModeActiveAction);
    hardware.getSceneButton(ButtonIndexes.Stop).getButton().releasedAction().addBinding(stopModeInactiveAction);

    HardwareActionBindable recordModeActiveAction = host.createAction(() -> setRecordMode(true), () -> "Record Mode Active");
    HardwareActionBindable recordModeInactiveAction = host.createAction(() -> setRecordMode(false), () -> "Record Mode Inactive");
    hardware.getSceneButton(ButtonIndexes.Record).getButton().pressedAction().addBinding(recordModeActiveAction);
    hardware.getSceneButton(ButtonIndexes.Record).getButton().releasedAction().addBinding(recordModeInactiveAction);

    //Delete Mode
    HardwareActionBindable deleteModeActiveAction = host.createAction(() -> setDeleteMode(true), () -> "Delete Mode Active");
    HardwareActionBindable deleteModeInactiveAction = host.createAction(() -> setDeleteMode(false), () -> "Delete Mode Inactive");
    hardware.getSceneButton(ButtonIndexes.Delete).getButton().pressedAction().addBinding(deleteModeActiveAction);
    hardware.getSceneButton(ButtonIndexes.Delete).getButton().releasedAction().addBinding(deleteModeInactiveAction);
  }

  private void setDeleteMode(boolean b) {
    deleteModeActive = b;
    updateState();
  }

  private void setRecordMode(boolean b) {
    recordModeActive = b;
    updateState();
  }

  private void setStopMode(boolean bool) {
    stopModeActive = bool;
    updateState();
  }

  private void setSceneLaunchMode(boolean active) {
    launchSceneModeActive = active;
    updateState();
  }

  private void updateState() {
    if (stopModeActive) {
      if (altLaunchModeActive) {
        bindStop();
        doubleGridState = DoubleGridState.StopAltMode;
      } else {
        bindStopAlt();
        doubleGridState = DoubleGridState.StopMode;
      }
    } else if (deleteModeActive) {
      bindActionsToGrid(ActionID.ClipDelete, null);
      doubleGridState = DoubleGridState.DeleteMode;
    } else if (launchSceneModeActive) {
      if (altLaunchModeActive) {
        bindSceneAltButtonsClips();
        doubleGridState = DoubleGridState.SceneAltMode;
      } else {
        bindSceneButtonsClips();
        doubleGridState = DoubleGridState.SceneMode;
      }
    } else if (recordModeActive){
      if (altLaunchModeActive) {
        bindRecordMode();
        doubleGridState = DoubleGridState.RecordAltMode;
      } else {
        bindRecordMode();
        doubleGridState = DoubleGridState.RecordMode;
    }
    } else {
      if (altLaunchModeActive) {
        bindLaunchAlt();
        doubleGridState = DoubleGridState.LaunchAltMode;
      } else {
        bindLaunch();
        doubleGridState = DoubleGridState.LaunchMode;
      }
    }

    //Update State Observers
    gridStatus.forEach(track -> track.forEach(gridStatus -> gridStatus.setGridState(doubleGridState)));

    if(launchSceneModeActive) {
      hardware.getSceneButton(ButtonIndexes.LaunchScene).setState(GridButtonColor.SceneModeActive);
    } else {
      hardware.getSceneButton(ButtonIndexes.LaunchScene).setState(GridButtonColor.SceneModeInactive);
    }
    if (stopModeActive) {
      hardware.getSceneButton(ButtonIndexes.Stop).setState(GridButtonColor.StopModeActive);
    } else {
      hardware.getSceneButton(ButtonIndexes.Stop).setState(GridButtonColor.StopModeInactive);
    }

    if (altLaunchModeActive) {
      hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonColor.LaunchAltMode);
    } else {
      hardware.getSceneButton(ButtonIndexes.LaunchMode).setState(GridButtonColor.LaunchNormalMode);
    }

    if(recordModeActive) {
      hardware.getSceneButton(ButtonIndexes.Record).setState(GridButtonColor.NoteRecording);
    } else {
      hardware.getSceneButton(ButtonIndexes.Record).setState(GridButtonColor.NoteOffRecording);
    }

    if(deleteModeActive) {
      hardware.getSceneButton(ButtonIndexes.Delete).setState(GridButtonColor.DeleteClipPlaying);
    } else {
      hardware.getSceneButton(ButtonIndexes.Delete).setState(GridButtonColor.DeleteHasClip);
    }

  }

  private void toggleLaunchModeState() {
    altLaunchModeActive = !altLaunchModeActive;
    updateState();
  }

  private int[] buttonMap(int col, int row) {
    if (col >= 8) {
      row += 4;
    }
    return new int[]{col % 8, row};
  }

  private void createButtonBindings() {
    gridStatus = new ArrayList<>(8);
    for (int trackIndex = 0; trackIndex < cursorTrackList.size(); trackIndex++) {
      CursorTrack cursorTrack = cursorTrackList.get(trackIndex);
      ClipLauncherSlotBank clipLauncherSlotBank = clipLauncherSlotBanks.get(trackIndex);
      clipLauncherSlotBank.scrollPosition().markInterested();
      int sizeOfBank = clipLauncherSlotBank.getSizeOfBank();
      gridStatus.add(new ArrayList<>(sizeOfBank));

      for (int rowIndex = 0; rowIndex < sizeOfBank; rowIndex++) {
        int[] mappedButton = buttonMap(trackIndex, rowIndex);
        int col = mappedButton[0];
        int row = mappedButton[1];
        ClipLauncherSlot clipLauncherSlot = clipLauncherSlotBank.getItemAt(rowIndex);

        //Map Actions
        Map<ActionID, HardwareActionBindable> actions = new HashMap<>();
        actions.put(ActionID.ClipLaunch, clipLauncherSlot.launchAction());
        actions.put(ActionID.ClipLaunchRelease, clipLauncherSlot.launchReleaseAction());
        actions.put(ActionID.ClipAltLaunch,  clipLauncherSlot.launchAltAction());
        actions.put(ActionID.ClipAltLaunchRelease, clipLauncherSlot.launchReleaseAltAction());
        actions.put(ActionID.ClipStop, clipLauncherSlotBank.stopAction());
        actions.put(ActionID.ClipAltStop, clipLauncherSlotBank.stopAltAction());
        //Scene
        Scene scene = sceneLauncher.getScene(rowIndex);
        actions.put(ActionID.SceneLaunch, scene.sceneLaunch());
        actions.put(ActionID.SceneAltLaunch, scene.sceneAltLaunch());
        actions.put(ActionID.SceneLaunchRelease, scene.launchSceneRelease());
        actions.put(ActionID.SceneAltLaunchRelease, scene.launchAltSceneRelease());
        actions.put(ActionID.RecordMode, clipLauncherSlot.recordAction());
        actions.put(ActionID.ClipDelete, clipLauncherSlot.deleteObjectAction());

        clipButtonActions[col][row] = actions;

        GridButton gridButton = hardware.getGridButton(col, row);

        gridStatus.get(trackIndex).add(new GridStatus(hardware.getGridButton(col, row), clipLauncherSlot, cursorTrack));
      }
    }
  }

  private void bindLaunch() {
    bindActionsToGrid(ActionID.ClipLaunch, ActionID.ClipLaunchRelease);
  }

  private void bindLaunchAlt() {
    bindActionsToGrid(ActionID.ClipAltLaunch, ActionID.ClipAltLaunchRelease);
  }
  private void bindStop() {
    bindActionsToGrid(ActionID.ClipStop, null);
  }

  private void bindStopAlt() {
    bindActionsToGrid(ActionID.ClipAltStop, null);
  }

  private void bindSceneButtonsClips() {
    bindActionsToGrid(ActionID.SceneLaunch, ActionID.SceneLaunchRelease);
  }

  private void bindSceneAltButtonsClips() {
    bindActionsToGrid(ActionID.SceneAltLaunch, ActionID.SceneAltLaunchRelease);
  }

  private void bindStopButtons() {
    bindActionsToGrid(ActionID.ClipStop, null);
  }

  private void bindRecordMode() {
    bindActionsToGrid(ActionID.RecordMode, ActionID.ClipLaunchRelease);
  }


  private void bindActionsToGrid(ActionID pressedAction, ActionID releasedAction) {
    for (int col = 0; col < 8; col++) {
      for (int row = 0; row < 8; row++) {
        bindActionsToGridButton(col, row, pressedAction, releasedAction);
      }
    }
  }

  private void bindActionsToGridButton(int col, int row, ActionID pressedAction, ActionID releasedAction) {
    Map<ActionID, HardwareActionBindable> actions = clipButtonActions[col][row];
    if (pressedAction != null) {
      hardware.getGridButton(col, row).getButton().pressedAction().setBinding(actions.get(pressedAction));
    } else {
      hardware.getGridButton(col, row).getButton().pressedAction().clearBindings();
    }
    if (releasedAction != null) {
      hardware.getGridButton(col, row).getButton().releasedAction().setBinding(actions.get(releasedAction));
    } else {
      hardware.getGridButton(col, row).getButton().releasedAction().clearBindings();
    }
  }

  public void moveClipLauncher(int direction) {
    int steps = direction * 4;
    this.sceneBank.scrollPosition().set(this.sceneBank.scrollPosition().get() + steps);
    for (int trackIndex = 0; trackIndex < cursorTrackList.size(); trackIndex++) {
      ClipLauncherSlotBank clipLauncherSlotBank = clipLauncherSlotBanks.get(trackIndex);
      int index = clipLauncherSlotBank.scrollPosition().get();
      clipLauncherSlotBank.scrollPosition().set(index + steps);

    }
  }

  // Scene Button Indexes
  private static class ButtonIndexes {
    static final int FillActive = 0;
    static final int FillAlt = 1;
    static final int LaunchMode = 2;
    static final int LaunchScene = 3;

    static final int Stop = 4;
    static final int Record = 5;
    static final int Delete = 6;
  }

  // ActionID Enum
  public enum ActionID {
    ClipLaunch,
    ClipLaunchRelease,
    ClipAltLaunch,
    ClipAltLaunchRelease,
    SceneLaunch,
    SceneLaunchRelease,
    SceneAltLaunch,
    SceneAltLaunchRelease,
    ClipStop,
    ClipAltStop,
    ClipDelete,
    RecordMode;
  }
}
